### PR TITLE
bug: Restrict user data access in User query

### LINKED ISSUE #2214: Unrestricted Exposure of User Data Through the User Query
**Describe the bug**
The affected query is `User($id: ID!)`  which fetches detailed information about a user based on the user ID provided. The current implementation **does not restrict users** from querying information about others, which violates common privacy principles and data access controls.

```
query User($id: ID!) {
    user(id: $id) {
        user {
            _id
            joinedOrganizations {
                _id
            }
            phone {
                mobile
            }
            address {
                line1
                countryCode
                city
                state
            }
        }
        appUserProfile {
            _id
            adminFor {
                _id
            }
        }
    }
}
```

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by nitintwt**
**What kind of change does this PR introduce?**

Implement user access control for querying user data.

**Issue Number:**

Fixes #2214

**Did you add tests for your changes?**

No

**Snapshots/Videos:**

**If relevant, did you update the documentation?**

No

**Summary**

This PR introduces user access control to ensure that users can only query their own profile data, addressing privacy concerns and data access violations. Previously, users could query information about other users, which goes against common privacy principles. This change mitigates the risk of unauthorized data access by checking the user ID before returning user information.

**Does this PR introduce a breaking change?**

No, this PR does not introduce any breaking changes.

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

Yes

## Summary by CodeRabbit

- **New Features**
	- Enhanced user data access control, ensuring users can only access their own profiles or those of users they are authorized to view (admins and super admins).
	- Implemented error handling for unauthorized access attempts, with specific error messages for invalid user IDs and access denials.

- **Bug Fixes**
	- Fixed potential issues with user data retrieval by ensuring proper user validation checks are in place.

- **Tests**
	- Introduced comprehensive unit tests for the user query resolver, validating scenarios for invalid user IDs, unauthorized access, admin and super admin access, and successful data retrieval for user profiles.

### CODE REVIEW

File: src/resolvers/Query/user.ts [hunk 1]
```diff
@@ -23,9 +27,18 @@ export const user: QueryResolvers["user"] = async (_parent, args, context) => {
...
+  if (context.userId !== args.id) {
+    throw new errors.UnauthorizedError(
+      "Access denied. You can only view your own profile.",
+    );
+  }
+
+  //Fetch the user data from the database based on the provided ID (args.id)
+  const user: InterfaceUser = (await User.findById(
+    args.id,
+  ).lean()) as InterfaceUser;
```
[the remark this answers is not in the document]
Comment (nitintwt) [reply src/resolvers/Query/user.ts:40]: The values of args.id and context.userId are the same, ensuring that users can only access their own information. I have already verified that context.userId exists; if it does not, an appropriate error will be thrown, which effectively prevents unauthorized access to user data.

**Review by palisadoes — CHANGES_REQUESTED** [review]
1. Please write tests so that this never happens again
2. Make sure that coderabbit approves the PR. Follow all its recommended changes.

**Review by palisadoes — DISMISSED** [review]

**Comment by nitintwt** [comment]
@pranshugupta54  Only admins of the specific organization in which the user is a member can access that user's data. Am I right ?

**Comment by nitintwt** [comment]
@pranshugupta54 SuperAdmin means the creator of the org ? In the organization schema , creatorId is SuperAdmin ?

**Comment by palisadoes** [comment]
> @pranshugupta54 SuperAdmin means the creator of the org ? In the organization schema , creatorId is SuperAdmin ?

https://docs.talawa.io/docs/introduction/core-concepts

**Comment by nitintwt** [comment]
The error says the source and target branches should be different. According to the docs, I need to use the develop branch locally for contributions. So, to resolve the failed check, I have to create a branch from the develop branch and then push it. However, I've already created the PR and can’t change the branch of my local repo. I’ll need to create a new PR to resolve this.

@pranshugupta54  Am I correct?

**Comment by pranshugupta54** [comment]
Yes, please open a new PR. Ensure the develop branch stays synced with the original repository, and create separate branches for any fixes.

Also, please reference this PR in your new one with a brief explanation for clarity.

**Pull request closed without merging**